core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -86,191 +88,117 @@ public Ec2Metadata.Builder toBuilder() {
+     * Gets the specified instance metadata value by the given path. Will retry base on the {@link Ec2MetadataRetryPolicy retry
+     * policy} provided, in the case of an IOException during request. Will not retry on SdkClientException, like 4XX/5XX HTTP
+     * error.
+     * @param path  Input path of the resource to get.
+     * @throws SdkClientException if the request for a token or the request for the Metadata does not have a 2XX SUCCESS response,
+     *                            if the maximum number of retries is reached,
+     *                            or if another IOException is thrown during the request.
      * @return Instance metadata value as part of MetadataResponse Object
      */
     @Override
     public MetadataResponse get(String path) {
-
-        MetadataResponse metadataResponse = null;
-        AbortableInputStream abortableInputStream = null;
-
-        for (int tries = 1 ; tries <= retryPolicy.numRetries() + 1; tries ++) {
-
+        Throwable lastCause = null;
+        for (int tries = 0; tries < retryPolicy.numRetries(); tries++) {
+            RetryPolicyContext retryPolicyContext = RetryPolicyContext.builder().retriesAttempted(tries).build();
             try {
-                Optional<String> token = getToken();
-                if (token.isPresent()) {
-                    HttpExecuteResponse response = getDataHttpResponse(path, token.get());
-
-                    int statusCode = response.httpResponse().statusCode();
-                    Optional<AbortableInputStream> responseBody = response.responseBody();
-
-                    if (statusCode == 200) {
-                        if (!responseBody.isPresent()) {
-                            throw SdkClientException.builder()
-                                                     .message("Response body empty with Status Code "  + statusCode).build();
-                        }
-                        abortableInputStream = responseBody.get();
-                        String data = IoUtils.toUtf8String(abortableInputStream);
-                        metadataResponse = new MetadataResponse(data);
-                        return metadataResponse;
-                    }
-                    handleException(statusCode, path);
-                }
-                //TODO Create IMDS Custom Exception
-            } catch (IOException io) {
-                log.warn(() -> "Received an IOException ", io);
-            } finally {
-                IoUtils.closeQuietly(abortableInputStream, log.logger());
+                String token = getToken();
+                return sendRequest(path, token);
+            } catch (SdkClientException sdkClientException) {
+                String msg = String.format("Error while executing EC2Metadata request. Total attempts: %d. %s",
+                                           tries + 1,
+                                           sdkClientException.getMessage());
+                log.debug(() -> msg);
```

Comment:
We should probably either create a variable `Supplier<String>` and pass it to `log.debug(() -> msgSupplier)` or just inline it `log.debug(() -> String.format("Error while executing EC2Metadata request. Total attempts: %d. %s", tries + 1, sdkClientException.getMessage())` to lazily create the string.

Same as line 121 and line 125

Reply:
ack